### PR TITLE
p2p: require all addresses come with an ID no matter what

### PR DESCRIPTION
**Pull request by melekes**
Refs #1228 

* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG.md

### CODE REVIEW

**Comment by ebuchman** [comment]
Looks like https://github.com/tendermint/tendermint/issues/1345 is rearing it's head. Maybe @xla can help work some magic on that one :P

File: p2p/netaddress.go [hunk 1]
```diff
@@ -49,33 +49,45 @@ func NewNetAddress(id ID, addr net.Addr) *NetAddress {
...
+	na.ID = id
+	return na
 }
 
-// NewNetAddressString returns a new NetAddress using the provided
-// address in the form of "ID@IP:Port", where the ID is optional.
+// NewNetAddressStringWithIDRequired returns a new NetAddress using the provided
+// address in the form of "ID@IP:Port".
 // Also resolves the host if host is not an IP.
 func NewNetAddressString(addr string) (*NetAddress, error) {
```
Comment (odeke-em) [inline p2p/netaddress.go]: Did you forget to change the name or should that name doc be reverted to `NewNetAddressString` from `NewNetAddressStringWithIDRequired`?

File: p2p/netaddress.go [hunk 2]
```diff
@@ -49,33 +49,45 @@ func NewNetAddress(id ID, addr net.Addr) *NetAddress {
...
 
 	var id ID
-	spl := strings.Split(addr, "@")
+	spl := strings.Split(addrWithoutProtocol, "@")
 	if len(spl) == 2 {
 		idStr := spl[0]
 		idBytes, err := hex.DecodeString(idStr)
 		if err != nil {
-			return nil, errors.Wrap(err, fmt.Sprintf("Address (%s) contains invalid ID", addr))
+			return nil, errors.Wrapf(err, "Address (%s) contains invalid ID", addrWithoutProtocol)
```
Comment (odeke-em) [inline p2p/netaddress.go]: If I recall right, `github.com/pkg/errors` is no longer recommended in Tendermint /cc @jaekwon @ebuchman. So, perhaps
```
fmt.Errorf("%v Address (%s) contains invalid ID", err, addrWithoutProtocol)
```

File: p2p/netaddress_test.go [hunk 3]
```diff
@@ -70,23 +90,21 @@ func TestNewNetAddressString(t *testing.T) {
 }
 
 func TestNewNetAddressStrings(t *testing.T) {
-	addrs, errs := NewNetAddressStrings([]string{"127.0.0.1:8080", "127.0.0.2:8080"})
+	addrs, errs := NewNetAddressStrings([]string{
```
Comment (odeke-em) [inline p2p/netaddress_test.go]: Could we also test some invalid addresses? The current test just confirms the always expected case.

File: p2p/netaddress.go (hunk 1 above)
Comment (melekes) [reply p2p/netaddress.go]: yep, thanks

File: p2p/netaddress.go (hunk 2 above)
Comment (melekes) [reply p2p/netaddress.go]: new lib is almost ready https://github.com/tendermint/tmlibs/pull/180. but let's not wait here please

Comment (odeke-em) [reply p2p/netaddress.go]: Cool cool, that was fast with how that new errors package was started and it is almost complete.

File: p2p/netaddress.go [hunk 4]
```diff
@@ -49,33 +49,45 @@ func NewNetAddress(id ID, addr net.Addr) *NetAddress {
...
 func NewNetAddressString(addr string) (*NetAddress, error) {
-	addr = removeProtocolIfDefined(addr)
+	spl := strings.Split(addr, "@")
+	if len(spl) < 2 {
+		return nil, fmt.Errorf("Address (%s) does not contain ID", addr)
+	}
+	return NewNetAddressStringWithOptionalID(addr)
+}
+
+// NewNetAddressStringWithIDRequired returns a new NetAddress using the provided
```
Comment (odeke-em) [inline p2p/netaddress.go]: Minor nit, the name in the doc doesn't match the actual function name:
- NewNetAddressStringWithIDRequired
+ NewNetAddressStringWithOptionalID

**Pull request merged**